src/core/Merger.cpp

```diff
@@ -0,0 +1,520 @@
+            targetChildGroup->setIcon(sourceChildGroup->iconUuid());
+        } else {
+            targetChildGroup->setIcon(sourceChildGroup->iconNumber());
+        }
+        targetChildGroup->setExpiryTime(sourceChildGroup->timeInfo().expiryTime());
+        // TODO HNH: Since we are updating our own group from the source group, I think we should update the timestamp
+    }
+    return changes;
+}
+
+bool Merger::markOlderEntry(Entry* entry)
+{
+    entry->attributes()->set(
+        "merged", tr("older entry merged from database \"%1\"").arg(entry->group()->database()->metadata()->name()));
+    return true;
+}
+
+void Merger::moveEntry(Entry* entry, Group* targetGroup)
+{
+    Q_ASSERT(entry);
+    Group* sourceGroup = entry->group();
+    if (sourceGroup == targetGroup) {
+        return;
+    }
+    const bool sourceGroupUpdateTimeInfo = sourceGroup ? sourceGroup->canUpdateTimeinfo() : false;
+    if (sourceGroup) {
+        sourceGroup->setUpdateTimeinfo(false);
+    }
+    const bool targetGroupUpdateTimeInfo = targetGroup ? targetGroup->canUpdateTimeinfo() : false;
+    if (targetGroup) {
+        targetGroup->setUpdateTimeinfo(false);
+    }
+    const bool entryUpdateTimeInfo = entry->canUpdateTimeinfo();
+    entry->setUpdateTimeinfo(false);
+
+    entry->setGroup(targetGroup);
+
+    entry->setUpdateTimeinfo(entryUpdateTimeInfo);
+    if (targetGroup) {
+        targetGroup->setUpdateTimeinfo(targetGroupUpdateTimeInfo);
+    }
+    if (sourceGroup) {
+        sourceGroup->setUpdateTimeinfo(sourceGroupUpdateTimeInfo);
+    }
+}
+
+void Merger::moveGroup(Group* group, Group* targetGroup)
+{
+    Q_ASSERT(group);
+    Group* sourceGroup = group->parentGroup();
+    if (sourceGroup == targetGroup) {
+        return;
+    }
+    const bool sourceGroupUpdateTimeInfo = sourceGroup ? sourceGroup->canUpdateTimeinfo() : false;
+    if (sourceGroup) {
+        sourceGroup->setUpdateTimeinfo(false);
+    }
+    const bool targetGroupUpdateTimeInfo = targetGroup ? targetGroup->canUpdateTimeinfo() : false;
+    if (targetGroup) {
+        targetGroup->setUpdateTimeinfo(false);
+    }
+    const bool groupUpdateTimeInfo = group->canUpdateTimeinfo();
+    group->setUpdateTimeinfo(false);
+
+    group->setParent(targetGroup);
+
+    group->setUpdateTimeinfo(groupUpdateTimeInfo);
+    if (targetGroup) {
+        targetGroup->setUpdateTimeinfo(targetGroupUpdateTimeInfo);
+    }
+    if (sourceGroup) {
+        sourceGroup->setUpdateTimeinfo(sourceGroupUpdateTimeInfo);
+    }
+}
+
+void Merger::eraseEntry(Entry* entry)
+{
+    Database* database = entry->database();
+    // most simple method to remove an item from DeletedObjects :(
+    const QList<DeletedObject> deletions = database->deletedObjects();
+    Group* parentGroup = entry->group();
+    const bool groupUpdateTimeInfo = parentGroup ? parentGroup->canUpdateTimeinfo() : false;
+    if (parentGroup) {
+        parentGroup->setUpdateTimeinfo(false);
+    }
+    delete entry;
```

Comment:
I don't think that this will help. The problem is part of the api for which I couldn't provide a better alternative. To use a smart pointer not just in a symbolic way, we would need to extract the entry itself wrapped in a smart pointer from the database - but this raises the question of responsibility for marking an object as deleted. Currently, there is no way to remove an entry from a database without deleting the actual runtime object. The `Merger` has to take care to unify the changes (including the timestamps) without introducing unexpected timestamp changes during the merge. This is ensured by moving every change operation (in this case the `delete`) in `setUpdateTimeinfo` brackets.  Therefore I would keep the `delete` since the entry is only accessible within the `Merger`.

Reply:
What I mean is: you are calling `delete entry` unconditionally. Instead of calling delete explicitly, you can also wrap the pointer in a QScopedPointer, which does that automatically when the scope ends. The behaviour is exactly the same, but it prevents you from forgetting to delete the entry if you later add a branch before it, for example.

Reply:
To use `QScopedPointer` I would need to write something like the following code:

```
if (parentGroup) {
    parentGroup->setUpdateTimeinfo(false);
}
{
    QScopedPointer<Entry> pointer(entry); 
}
if (parentGroup) {
    parentGroup->setUpdateTimeinfo(...);
}
```
since it is necessary to delete the entry at this exact point in code. Using a smart pointer, I need to wrap it into an scope which is cleared immediately or the `setUpdateTimeinfo` bracket would be rendered useless because the deletion of entry would modify the database (write `DeletedObjects` and add change markers which were not part of the source databases).

Reply:
I see. You are making use of the Entry destructor, which manages removal from the group. I'm not too happy about Entries managing their parents, but oh well... lots of things to improve. Then I guess leave it as is for now.